src/pathogena/util.py

```diff
@@ -271,71 +280,574 @@ def hash_file(file_path: Path) -> str:
+    upload session.
+
+    Args:
+        upload_data (UploadFileType): An object containing the upload configuration,
+            including the batch ID, access token, environment, and file details.
+        instrument_code (str): The instrument code used to generate the data. This is
+            passed to the backend API for processing.
+        api_client (APIClient): Instance of the APIClient class.
+        sample_uploads (dict[str, Any] | None): State of sample uploads, if available.
+
+    Returns:
+        None
+    """
+    file_status = {}
+
+    # prepare files before uploading
+    file_preparation = prepare_files(
+        batch_pk=upload_data.batch_pk,
+        instrument_code=instrument_code,
+        files=upload_data.files,
+        api_client=api_client,
+        sample_uploads=sample_uploads,
+    )
+
+    # handle any errors during preparation
+    if "API error occurred" in file_preparation:
+        logging.error(
+            f"Error preparing files: {file_preparation['API error occurred']}"
+        )
+        return file_preparation
+
+    # files have been sucessfully prepared, extract the prepared file list
+    selected_files = file_preparation["files"]
+
+    # upload the file chunks
+    with ThreadPoolExecutor(max_workers=upload_data.max_concurrent_chunks) as executor:
+        futures = []
+        for file in selected_files:
+            file = file["file"]
+            future = executor.submit(upload_chunks, upload_data, file, file_status)
+            futures.append(future)
+
+        for future in as_completed(futures):
+            try:
+                future.result()
+            except Exception as e:
+                logging.error(f"Error uploading file: {e}")
+
+    # end the upload session
+    end_session = APIClient.batches_samples_end_upload_session_create(
+        upload_data.batch_pk, upload_data
+    )
+
+    if end_session["status"] != 200:
+        logging.error(f"Failed to end upload session for batch {upload_data.batch_id}.")
+    else:
+        logging.info(f"All uploads complete.")
+
+
+def end_upload(batch_pk: int, file_id: str) -> dict[str, Any]:
+    """End the upload of a sample.
+
+    Args:
+        batch_pk (int):ID of the uploaded sample.
+        upload_session (int): ID of the upload session.
+
+    Returns:
+        dict[str, Any]: The response JSON from the API.
+    """
+    data = {"upload_id": file_id}
+
+    return APIClient.batches_uploads_end_create(batch_pk, data)
+
+
+def end_upload_session(batch_pk: int, upload_session: int) -> dict[str, Any]:
+    """End the upload session.
+
+    Args:
+        batch_pk (int):ID of the uploaded sample.
+        upload_session (int): ID of the upload session.
+
+    Returns:
+        dict[str, Any]: The response JSON from the API.
+    """
+    data = {"upload_session": upload_session}
+    return APIClient.batches_samples_end_upload_session_create(batch_pk, data)
 
 
 def upload_fastq(
-    sample_id: int,
+    batch_pk: int,
```

Comment:
we need to update parameters of upload_fastq() calls

1. https://github.com/EIT-Pathogena/client/blob/e7710c6fd7ddd9ab1e53d03631ae29e83a02c775/src/pathogena/lib.py#L413
2. https://github.com/EIT-Pathogena/client/blob/e7710c6fd7ddd9ab1e53d03631ae29e83a02c775/src/pathogena/lib.py#L422